target/linux/bcm53xx/patches-5.10/342-Meraki-MX6X-Status-LEDs.patch

```diff
@@ -0,0 +1,47 @@
+From: Matthew Hagan <mnhagan88@gmail.com>
+Date: Wed, 17 Mar 2021 10:47:39 +0000
+Subject: bcm-nsp: add MX64/MX65 LED aliases
+
+Add indicator LEDs to the MX64/MX65 device trees.
```

Comment:
Why is that change necessary?

Note that color/function setup does not work with our DT aliases at the moment, we need "label" property.

Reply:
It seems with an LED to function LED_FUNCTION_POWER, it will be used - at least providing "led-boot" and "led-running" functionality. I'm not sure if this is intended behaviour or why this happens - possibly related to the "green:power" LED name generated in /sys/class/leds? If you prefer I can patch the DT in the mean time to use the label property on all LEDs. This would also provide a better description for the Ethernet LEDs.

Reply:
Unfortunately, the DT aliases led-boot etc. do not work with the color/function scheme for LEDs. The LED DT node names the kernel generates are not fully predictable and there are other issues with their new approach as well. The full story may be found in #3437 .

I've invested quite a lot of time into getting OpenWrt ready for this and was quite frustrated to see that the new system is unusable for us. At least it was back then; the maintainers more or less openly told that they actually hadn't thought through the design completely there.

If you want to have support for the aliases, just put an additional label property into the DTS. That should be easy enough for the near future at least.

Reply:
FYI bcm53xx target uses a [custom diag.sh](https://github.com/openwrt/openwrt/blob/master/target/linux/bcm53xx/base-files/etc/diag.sh#L15) which sets an LED with "power" label to status LED before seeking any dt-aliases.

Reply:
Oh, thanks, haven't really checked that. That explains to observation above.

Still, I don't think that changing the LED function (name) in order to use an ancient workaround should be the way to go for new devices.

Reply:
I have reverted all LEDs to use the legacy bindings. Until the LED subsystem is resolved and we can provide more detail, whether by prepending device information or something else, I don't see any other way forward.

By the way @chunkeey that diag.sh means that [331-Meraki-MR32-Status-LEDs.patch](https://github.com/openwrt/openwrt/blob/master/target/linux/bcm53xx/patches-5.10/331-Meraki-MR32-Status-LEDs.patch) isn't doing anything useful, since the status is inferred though the name provided by `LED_FUNCTION_POWER`.

Reply:
> By the way @chunkeey doesn't that diag.sh mean that [331-Meraki-MR32-Status-LEDs.patch](https://github.com/openwrt/openwrt/blob/master/target/linux/bcm53xx/patches-5.10/331-Meraki-MR32-Status-LEDs.patch) isn't doing anything useful, since the status LED is inferred though the name provided by `LED_FUNCTION_POWER`?

Last two lines of get_status():
 
    # And finally, let's also try the device-Tree aliases node
    status_led="$(get_dt_led status)"

so, this is where the status LED is wired up.

Reply:
My understanding in this case is that the LED node will be evaluated by the subsystem to "green:power" and as a result line 9 of get_status_led() will trigger and return, so the last two lines won't be evaluated. To achieve the same effect I had changed the function of the same LED on the MX devices to `LED_FUNCTION_POWER`, the alias wasn't needed to make it work. Changing this LED to something else like `LED_FUNCTION_INDICATOR` would stop it from working. To make it work with aliases, as @adschm mentioned, the label property must be used.

Reply:
Ah, these function + color properties were there for decoration. In the context of the leds-pwm driver/compatible, these are all ignored.

https://github.com/torvalds/linux/blob/master/drivers/leds/leds-pwm.c#L143 
(apart from of_ to fwnode, this is pretty much the same from back then)

This is where leds-pwm iterates over/parses the sub-nodes. And as you can see, it only cares about the name only work if there was a custom "label" property. So that's why the 331- "does something useful" in the meantime, until either led-pwm learns to deal with the color+function properties or it's replacement. (You see what I did there ;-) )